qemu, runInLinuxVM: fix KVM availability check

mschwaig opened the pull request:
KVM should only be considered abailable if /dev/kvm exists and
is read-writable by the user that is trying to launch it.

The previous check for existance only had the consequence that
on some Linux distributions running VMs with Nix's QEMU only worked
if KVM was NOT installed.

See issue: https://github.com/NixOS/nixpkgs/issues/124371

---

To test this change I
1. built some NixOS Test driver on `x86_64` NixOS.
2. changed the permissions for /dev/kvm `chmod o-rw /dev/kvm`
3. ran `result/bin/nixos-run-vms`

Without this change that test failed right away with a log like this:
```
./nixos-run-vms
starting VDE switch for network 1
deleting VM state directory /tmp/vm-state-pia
if you want to keep the VM state, pass --keep-vm-state
deleting VM state directory /tmp/vm-state-satellite
if you want to keep the VM state, pass --keep-vm-state
deleting VM state directory /tmp/vm-state-sensor
if you want to keep the VM state, pass --keep-vm-state
running the VM test script
starting all VMs
pia: starting vm
pia # Formatting '/tmp/vm-state-pia/pia.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib size=536870912 lazy_refcounts=off refcount_bits=16
pia # Could not access KVM kernel module: Permission denied
pia # qemu-system-x86_64: failed to initialize kvm: Permission denied
pia: QEMU running (pid 2926214)
satellite: starting vm
satellite # Formatting '/tmp/vm-state-satellite/satellite.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib size=536870912 lazy_refcounts=off refcount_bits=16
satellite # Could not access KVM kernel module: Permission denied
satellite # qemu-system-x86_64: failed to initialize kvm: Permission denied
satellite: QEMU running (pid 2926260)
sensor: starting vm
sensor # Formatting '/tmp/vm-state-sensor/sensor.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib size=536870912 lazy_refcounts=off refcount_bits=16
sensor # Could not access KVM kernel module: Permission denied
sensor # qemu-system-x86_64: failed to initialize kvm: Permission denied
sensor: QEMU running (pid 2926271)
(1.33 seconds)
waiting for all VMs to finish
pia: waiting for the VM to power off
(0.00 seconds)
satellite: waiting for the VM to power off
(0.00 seconds)
sensor: waiting for the VM to power off
(0.00 seconds)
(0.00 seconds)
(1.33 seconds)
test script finished in 1.33s
cleaning up
(0.00 seconds)
```
while with the change the test proceeded without hardware acceleration via KVM.

I have also tested that this fixes the same issue when running QEMU inside the build sandbox as part of a NixOS Test. The NixOS Test checkbox is unchecked though, because there there is automatic NixOS Tests for the new behavior.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mschwaig committed:
qemu, runInLinuxVM: fix KVM availability check

KVM should only be considered abailable if /dev/kvm exists and
is read-writable by the user that is trying to launch it.

The previous check for existance only had the consequence that
on some Linux distributions running VMs with Nix's QEMU only worked
if KVM was NOT installed.

fixes #124371

r-rmcgibbo commented:
Result of `nixpkgs-review pr 125948` at d22a9ecf run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>qemu_xen</li>
    <li>qemu_xen-light</li>
    <li>qemu_xen_4_10</li>
    <li>qemu_xen_4_10-light</li>
  </ul>
</details>
<details>
  <summary>2 packages skipped due to time constraints:</summary>
  <ul>
    <li>multibootusb</li>
    <li>qemu_full</li>
  </ul>
</details>
<details>
  <summary>18 packages built successfully:</summary>
  <ul>
    <li>alpine-make-vm-image</li>
    <li>aqemu</li>
    <li>cloud-init</li>
    <li>cloud-utils</li>
    <li>cot (python38Packages.cot)</li>
    <li>gnome.gnome-boxes</li>
    <li>libguestfs</li>
    <li>out-of-tree</li>
    <li>python38Packages.guestfs</li>
    <li>python39Packages.cot</li>
    <li>python39Packages.guestfs</li>
    <li>qemu</li>
    <li>qemu-utils</li>
    <li>qemu_kvm</li>
    <li>qemu_test</li>
    <li>qtemu</li>
    <li>solo5</li>
    <li>vagrant</li>
  </ul>
</details>

lukegb commented:
My gut feeling is that the wrapper should probably warn on stderr if KVM isn't available. It's a bit weird having a wrapper called "qemu-kvm" that doesn't actually use kvm, in any case, but that's the situation we're in :P

I guess `qemu-maybe-kvm-if-it-seems-available` is too long ;)

mschwaig commented:
I can try emitting a warning on stderr from the wrapper and check how visible that is later this week.

---
It seems the right component to make sure people know what's going on with KVM would actually be Nix and the handling of the `kvm` system feature there.

We could make sure it works and takes the permissions of the build user into account. This seems to be the code for detecting the system feature:
https://github.com/NixOS/nix/blob/7c90552879da4d1df99b50c85e94201981e60123/src/libstore/globals.cc#L128-L130

I have not tested this yet because somehow my VM without `/dev/kvm` also has the `kvm` feature.

I think the way to enable non-accelreated execution should be manually adding the `kvm` system feature. That's already how it normally works right now, execpt for the edge case addressed in this PR.

To make things more clear to users
- assuming the `kvm` system feature check works and
- determines you don't have a usable KVM

it could explicitly prompt you to either
- install KVM and set the permissions in a workable way so Nix picks it up or
- add it to the system features manually and accept the performance pentalty.

That way at least as part of a build things only take the slower route if you set some deliberate action after a prompt that informs you about the consequences.

(There is also some code which suggests that supplementary groups should actually stick to the build users here:
https://github.com/NixOS/nix/blob/b10256af51dfa929e8f916414d6f021dd45f2e1e/src/libstore/build/local-derivation-goal.cc#L1901-L1902, but from testing while looking into https://github.com/NixOS/nixpkgs/issues/124371 it seems that this is at least not working for all confirgurations of the build sandbox.)

I'm thinking about opening an issue about this whole `kvm` system feature topic on Nix if people think that makes sense.
Some relevant discussion about how that feature is handeled: https://github.com/NixOS/nix/issues/2964#issuecomment-504097120

puffnfresh reviewed (approved):
Definitely an improvement on the current situation

Artturin commented:
> I can try emitting a warning on stderr from the wrapper and check how visible that is later this week.

Planning on finishing this?

mschwaig committed:
qemu: emit warnings when KVM acceleration is not usable

This fixes the qemu-kvm wrapper we add for convenience
silently not using KVM, when the system would support it
by at least leaving an indication in the log that the build ran
slower because it ran without KVM.

mschwaig commented:
> > I can try emitting a warning on stderr from the wrapper and check how visible that is later this week.
> 
> Planning on finishing this?
> 
> --run could be useful here
> 
> https://github.com/NixOS/nixpkgs/blob/9fcf2a8a2cb0f5b78edc8ec7ca877240e7fe3009/pkgs/build-support/setup-hooks/make-wrapper.sh#L21

Thanks for the hint. I was unsure how to best approach this and got busy with other things and your comment was very helpful.

I added warnings now.

mschwaig commented:
It seems the slow path of using QEMU without KVM when `/dev/kvm` is not read-/writable is generally broken right now.

On my machine QEMU now gets stuck at
```
# Booting from ROM...
# Probing EDD (edd=off to disable)... ok
```
This was working fine yesterday but I can't reproduce what changed.
Everything's also working fine when KVM is available.

The following change fixes it.

```diff
diff --git a/nixos/lib/qemu-common.nix b/nixos/lib/qemu-common.nix
index 84f9060acd6..ebcdd61f587 100644
--- a/nixos/lib/qemu-common.nix
+++ b/nixos/lib/qemu-common.nix
@@ -22,7 +22,7 @@ rec {
         else throw "Unknown QEMU serial device for system '${pkgs.stdenv.hostPlatform.system}'";
 
   qemuBinary = qemuPkg: {
-    x86_64-linux = "${qemuPkg}/bin/qemu-kvm -cpu max";
+    x86_64-linux = "${qemuPkg}/bin/qemu-kvm -cpu qemu64";
     armv7l-linux = "${qemuPkg}/bin/qemu-system-arm -enable-kvm -machine virt -cpu host";
     aarch64-linux = "${qemuPkg}/bin/qemu-system-aarch64 -enable-kvm -machine virt,gic-version=host -cpu host";
     powerpc64le-linux = "${qemuPkg}/bin/qemu-system-ppc64 -machine powernv";
```
It's similar to what's reported in https://github.com/NixOS/nixpkgs/issues/141596#issue-1025938610

alyssais commented:
@mschwaig has anybody made a PR for that?  LGTM — I doubt we have many CPU-bound NixOS tests, and that's all qemu-common is used for AFAICT.

mschwaig commented:
It looks like `runInLinuxVM` and `make-disk-image.nix` use `qemu-common.nix`. Probably not CPU-bound but still I think access to KVM could make a significant difference there for performance there and for NixOS tests. I don't think a smaller set of available instructions would matter.

Would switching to `qemu64` disable KVM acceleration or just reduce what instructions are accelerated/supported?

I am  not aware of another PR that changes the emulated CPU.

If it just reduces the set of accelerated or supported instructions, I can add a commit for that here and I think we should merge this PR. Otherwise I'm not sure how to proceed.

mschwaig commented:
In the future I would also like to test this with your `qemu-6.2.0` branch @alyssais (https://github.com/NixOS/nixpkgs/pull/146526) to see if 6.2 will fix qemu getting stuck, so we could eventually revert to `-cpu max` in a future PR, but somehow I cannot fetch that branch from your fork right now.

mschwaig commented:
Working on this made me think it would make sense for a script that gets invoked before qemu to not only emit warnings but actually determine with which `-cpu` setting qemu gets invoked. For builds outside a VM the supported instruction has an impact on the output anyways, and `-max` already makes a machine-dependent determination of the emulated CPU based on the host CPU. If we made that determination ourselves we could
* write to the log what CPU was emulated,
* account for other factors, like access to `/dev/kvm` and
* tell qemu more specifically what to emulate.

The XML for describing CPUs and `host-model` concept from libvirt (as seen in [this presentation]( https://events19.linuxfoundation.org/wp-content/uploads/2017/12/Kashyap-Chamarthy_Effective-Virtual-CPU-Configuration-OSS-EU2018.pdf) look interesting for that.

alyssais commented:
> In the future I would also like to test this with your `qemu-6.2.0` branch @alyssais (https://github.com/NixOS/nixpkgs/pull/146526) to see if 6.2 will fix qemu getting stuck, so we could eventually revert to `-cpu max` in a future PR, but somehow I cannot fetch that branch from your fork right now.

That's very strange.  Does git fetch https://github.com/NixOS/nixpkgs pull/146526/head work?

alyssais commented:
> Working on this made me think it would make sense for a script that gets invoked before qemu to not only emit warnings but actually determine with which `-cpu` setting qemu gets invoked. For builds outside a VM the supported instruction has an impact on the output anyways, and `-max` already makes a machine-dependent determination of the emulated CPU based on the host CPU. If we made that determination ourselves we could
> * write to the log what CPU was emulated,
> * account for other factors, like access to `/dev/kvm` and
> * tell qemu more specifically what to emulate.
>
> The XML for describing CPUs and `host-model` concept from libvirt (as seen in [this presentation]( https://events19.linuxfoundation.org/wp-content/uploads/2017/12/Kashyap-Chamarthy_Effective-Virtual-CPU-Configuration-OSS-EU2018.pdf) look interesting for that.

I think qemu-kvm originates from Fedora (they certainly also add it).
Maybe it would be worth looking into if they do anything like this?

One of the purposes of qemu-kvm is compatibility with other
distributions that provide it, so I'd be hesitant to change it too much
from what those other distros do.

mschwaig commented:
> That's very strange. Does git fetch https://github.com/NixOS/nixpkgs pull/146526/head work?

Yes, that worked. Your commit from `qemu-6.2.0` (ded05e005c2f359d79cc88401c98699f0f92ecbd) prevents VMs from getting stuck on bootup when `/dev/kvm` has the wrong permissions, it seems qemu 6.2.0 fixes that issue.

Since the release date for qemu will be December 7 at the earliest I will add a commit here to change change `-cpu` to `qemu64` which can then be reverted as part of https://github.com/NixOS/nixpkgs/pull/146526.

That workaround should fix https://github.com/NixOS/nixpkgs/issues/141596 as well and hopefully qemu 6.2.0 fixes the actual problem there as well.

mschwaig committed:
qemu, runInLinuxVM: change default cpu to qemu64

The flag -cpu max leaves QEMU 6.1.0 stuck on some systems,
for example when /dev/kvm is not read-writable.
This does not happen with -cpu qemu64.

Getting stuck like that is a regression in 6.1.0 not yet present in 6.0.0
and should be fixed with 6.2.0 according to early testing with rc1.

We should consider reverting this change when we merge QEMU 6.2.0.
See #146526.

fixes #141596

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -222,7 +237,8 @@ stdenv.mkDerivation rec {
                   $out/bin/qemu-kvm \
-                  --add-flags "\$([ -e /dev/kvm ] && echo -enable-kvm)"
+                  --run ${emit-kvm-warnings}/bin/emit-kvm-warnings \
+                  --add-flags "\$([ -r /dev/kvm ] && [ -w /dev/kvm ] && echo -enable-kvm)"
```
Nitpick: no need for two invocations of `[` here.

```suggestion
                  --add-flags "\$([ -r /dev/kvm -a -w /dev/kvm ] && echo -enable-kvm)"
```

mschwaig replied:
Done in d033935ab6fd850160e31a1db822486f39650210

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -35,6 +35,21 @@ let
+    if [ ! -e /dev/kvm ]; then
+      echo -e "''${WARNING} KVM is not installed - execution will be slow"
+      echo "Consider installing KVM for hardware-accelerated execution."
+    elif [ ! -r /dev/kvm ] || [ ! -w /dev/kvm ]; then
```
Nitpick: no need for two invocations of `[` here.

```suggestion
    elif [ ! -r /dev/kvm -o ! -w /dev/kvm ]; then
```

mschwaig replied:
Done in d033935ab6fd850160e31a1db822486f39650210

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -35,6 +35,21 @@ let
     + lib.optionalString sdlSupport "sdl,";
+  emit-kvm-warnings = writeShellScriptBin "emit-kvm-warnings" ''
+    WARNCOL='\033[0m'
+    NEUTRALCOL='\033[0m'
```
I don't think these two definitions are supposed to be the same.

mschwaig replied:
Thanks, I don't know how that had happened. d033935ab6fd850160e31a1db822486f39650210

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -222,7 +237,8 @@ stdenv.mkDerivation rec {
       makeWrapper $out/bin/qemu-system-${stdenv.hostPlatform.qemuArch} \
                   $out/bin/qemu-kvm \
-                  --add-flags "\$([ -e /dev/kvm ] && echo -enable-kvm)"
+                  --run ${emit-kvm-warnings}/bin/emit-kvm-warnings \
```
Might it be better to have this file inside the QEMU derivation?  You could use `passAsFile` and then just copy it into $out/libexec or something.

mschwaig replied:
done in 88899a82dac3c78805b28c490259f1a4c3dc82f8

alyssais reviewed (changes requested):
The important change here is fixing the definition of `WARNCOL`, but I think everything else I've pointed out is also worth considering.

alyssais commented on nixos/lib/qemu-common.nix:
```diff
@@ -22,7 +22,7 @@ rec {
 
   qemuBinary = qemuPkg: {
-    x86_64-linux = "${qemuPkg}/bin/qemu-kvm -cpu max";
+    x86_64-linux = "${qemuPkg}/bin/qemu-kvm -cpu qemu";
```
I assume this is the same as `qemu64`?  If so, it's the default, so we could just omit the `-cpu`.

mschwaig replied:
I changed it to `qemu64`, I did not want to omit it. 3baacbf30f6b81af2b5bd3341da157093c0ba299

mschwaig commented:
> This commit message should start with "qemu: ".
 Rewritten as bdbb79178461c2e679bc30113a0bde575ab04c0f with that change.

mschwaig commented:
> The important change here is fixing the definition of `WARNCOL`, but I think everything else I've pointed out is also worth considering.

Thanks for the review. I have addressed everything except for not dropping the `-cpu` parameter, and I hope 88899a82dac3c78805b28c490259f1a4c3dc82f8 does what you meant with `passAsFile` since I had not used that feature before.

From comparing run times I can also confirm that hardware acceleration works still works with `-cpu qemu64`.

alyssais commented:
Looks good on a first glance, but I'm too tired to say for sure tonight and will have a closer look tomorrow.  Could you squash all the warning stuff into a single commit, and add appropriate "Fixes:" lines to your commit messages please?

mschwaig commented:
> Looks good on a first glance, but I'm too tired to say for sure tonight and will have a closer look tomorrow. Could you squash all the warning stuff into a single commit, and add appropriate "Fixes:" lines to your commit messages please?

I have reordered and squashed the commits so there is one commit per topic, added which issues they fix and moved the emitted warning from `stdout` to `stderr`.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -233,6 +236,25 @@ stdenv.mkDerivation rec {
+    NEUTRALCOL='\033[0m'
+    WARNING="''${WARNCOL}warning:''${NEUTRALCOL}"
+    if [ ! -e /dev/kvm ]; then
+      echo -e "''${WARNING} KVM is not installed - execution will be slow" >&2
```
This should be "loaded" or "available", not "installed", because the presence of KVM can be changed at runtime by inserting or removing the module.

mschwaig replied:
I have updated the wording to include both the module not being loaded and KVM not being installed.

alyssais reviewed (changes requested):
Sorry for the delay.  I've now tested this and am happy with it apart from this small terminology issue.

stephank commented:
These changes broke the Darwin build, but simply because there's no `libexec`: https://hydra.nixos.org/build/159925938/nixlog/1

```
cp: cannot create regular file '/nix/store/60s7gdxrrlyn8dm7d73v6r439v798kkc-qemu-6.1.0/libexec/emit-kvm-warnings': No such file or directory
```

However, I wonder what `qemu-kvm` should do at all on Darwin. If our tests depend on it, and we want to maybe support running tests on Darwin in the future, the wrapper should stay and do something useful, I think?

I believe `-enable-kvm` [is the same as](https://gitlab.com/qemu-project/qemu/-/blob/682aa69b1f4dd5f2905a94066fa4c853adc33251/softmmu/vl.c#L3348-3350) `-accel kvm`? Maybe the wrapper can do `-accel hvf` on Darwin? Regardless, the warning should be disabled on Darwin. Telling users to install KVM makes no sense.

Mic92 commented:
> These changes broke the Darwin build, but simply because there's no `libexec`: https://hydra.nixos.org/build/159925938/nixlog/1
> 
> ```
> cp: cannot create regular file '/nix/store/60s7gdxrrlyn8dm7d73v6r439v798kkc-qemu-6.1.0/libexec/emit-kvm-warnings': No such file or directory
> ```
> 
> However, I wonder what `qemu-kvm` should do at all on Darwin. If our tests depend on it, and we want to maybe support running tests on Darwin in the future, the wrapper should stay and do something useful, I think?
> 
> I believe `-enable-kvm` [is the same as](https://gitlab.com/qemu-project/qemu/-/blob/682aa69b1f4dd5f2905a94066fa4c853adc33251/softmmu/vl.c#L3348-3350) `-accel kvm`? Maybe the wrapper can do `-accel hvf` on Darwin? Regardless, the warning should be disabled on Darwin. Telling users to install KVM makes no sense.

see https://github.com/NixOS/nixpkgs/pull/148251

alyssais commented:
> Maybe the wrapper can do -accel hvf on Darwin? Regardless, the warning should be disabled on Darwin. Telling users to install KVM makes no sense.

This all sounds good.  I'd be happy with a PR for any or all of these things.

mschwaig commented:
I have opened #148305 which should remove the warning on other platforms, but was not able to test this change on Darwin myself.

I don't know how we would need to adapt the wrapper and warning be useful on for Darwin, but it sounds like a nice idea.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/runnixostest-is-faster-when-interactive/69220/4
